File: src/network/libqpassworddigestor.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const qcryptographichash_enums = @import("../libqcryptographichash.zig").enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qpassworddigestor.html
pub const qpassworddigestor = struct {
    /// [Qt documentation](https://doc.qt.io/qt-6/qpassworddigestor.html#deriveKeyPbkdf1)
    ///
    /// ``` param1: qcryptographichash_enums.Algorithm, param2: []u8, param3: []u8, param4: i32, param5: u64, allocator: std.mem.Allocator ```
    pub fn DeriveKeyPbkdf1(param1: i32, param2: []u8, param3: []u8, param4: i32, param5: u64, allocator: std.mem.Allocator) []u8 {
        const param2_str = qtc.libqt_string{
            .len = param2.len,
            .data = param2.ptr,
        };
        const param3_str = qtc.libqt_string{
            .len = param3.len,
            .data = param3.ptr,
        };
        const _bytearray: qtc.libqt_string = qtc.QPasswordDigestor_DeriveKeyPbkdf1(@intCast(param1), param2_str, param3_str, @intCast(param4), @intCast(param5));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qpassworddigestor.DeriveKeyPbkdf1: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qpassworddigestor.html#deriveKeyPbkdf2)
    ///
    /// ``` param1: qcryptographichash_enums.Algorithm, param2: []u8, param3: []u8, param4: i32, param5: u64, allocator: std.mem.Allocator ```
    pub fn DeriveKeyPbkdf2(param1: i32, param2: []u8, param3: []u8, param4: i32, param5: u64, allocator: std.mem.Allocator) []u8 {
        const param2_str = qtc.libqt_string{
            .len = param2.len,
            .data = param2.ptr,
        };
        const param3_str = qtc.libqt_string{
            .len = param3.len,
            .data = param3.ptr,
        };
        const _bytearray: qtc.libqt_string = qtc.QPasswordDigestor_DeriveKeyPbkdf2(@intCast(param1), param2_str, param3_str, @intCast(param4), @intCast(param5));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qpassworddigestor.DeriveKeyPbkdf2: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }
};

File: src/network/libqpassworddigestor.cpp
#include <QByteArray>
#include <QPasswordDigestor>
#include <qpassworddigestor.h>
#include "libqpassworddigestor.h"
#include "libqpassworddigestor.hxx"

libqt_string QPasswordDigestor_DeriveKeyPbkdf1(int param1, const libqt_string param2, const libqt_string param3, int param4, unsigned long long param5) {
    QByteArray param2_QByteArray(param2.data, param2.len);
    QByteArray param3_QByteArray(param3.data, param3.len);
    QByteArray _qb = QPasswordDigestor::deriveKeyPbkdf1(static_cast<QCryptographicHash::Algorithm>(param1), param2_QByteArray, param3_QByteArray, static_cast<int>(param4), static_cast<quint64>(param5));
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string QPasswordDigestor_DeriveKeyPbkdf2(int param1, const libqt_string param2, const libqt_string param3, int param4, unsigned long long param5) {
    QByteArray param2_QByteArray(param2.data, param2.len);
    QByteArray param3_QByteArray(param3.data, param3.len);
    QByteArray _qb = QPasswordDigestor::deriveKeyPbkdf2(static_cast<QCryptographicHash::Algorithm>(param1), param2_QByteArray, param3_QByteArray, static_cast<int>(param4), static_cast<quint64>(param5));
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}
